src/control/cmd/daos_agent/infocache.go

```diff
@@ -123,16 +158,19 @@ func (c *localFabricCache) Cache(ctx context.Context, scan []*netdetect.FabricSc
 	c.localNUMAFabric = NUMAFabricFromScan(ctx, c.log, scan, c.getDevAlias)
 
 	c.initialized.SetTrue()
-	return nil
 }
 
 // GetDevices fetches an appropriate fabric device from the cache.
 func (c *localFabricCache) GetDevice(numaNode int, netDevClass uint32) (*FabricInterface, error) {
 	if c == nil {
-		return nil, errors.New("nil localFabricCache")
+		return nil, NotCachedErr
 	}
+
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
 	if !c.IsCached() {
```

Comment:
this seems a little unintuitive to me, if data is not cached then I would expect a fetch

Reply:
I shifted the business logic outside the cache, so it's the caller who has to perform the fetch.